perun/collect/complexity/strategy.py

```diff
@@ -31,100 +35,139 @@ def get_default_strategy():
     """
 
+    # Remove duplicate rules, merge sampled / non-sampled rule lists and optionally pair the rules
+    kwargs['func'] = _remove_duplicate_probes(_merge_probes_lists(
+        func, func_sampled, kwargs['global_sampling']))
+    kwargs['static'] = _remove_duplicate_probes(_pair_rules(_merge_probes_lists(
+        static, static_sampled, kwargs['global_sampling'])))
+    kwargs['dynamic'] = _remove_duplicate_probes(_pair_rules(_merge_probes_lists(
+        dynamic, dynamic_sampled, kwargs['global_sampling'])))
+
     # Build sampling dictionary
-    samples = _build_samples(sampling, rules, global_sampling)
-
-    # Assembly the script
-    script = ''
-    # Add sampling counters
-    for _, sample in samples.items():
-        script += 'global samp_{0} = {1}\n'.format(str(sample[1]), sample[0] - 1)
-    script += '\n'
-
-    # Add probes
-    for rule in rules:
-        probe_in = 'probe process("{0}").function("{1}").call {{\n'.format(cmd, rule)
-        probe_out = 'probe process("{0}").function("{1}").return {{\n'.format(cmd, rule)
-        call = 'printf("%s %s\\n", thread_indent(1), probefunc())\n'
-        ret = 'printf("%s\\n", thread_indent(-1))\n'
-
-        if rule in samples:
-            # Probe should be sampled, add counter manipulation
-            probe_in += ('\tsamp_{0}++\n'
-                         '\tif(samp_{0} == {1}) {{\n'
-                         '\t\t{2}'
-                         '\t\tsamp_{0} = 0\n'
-                         '\t}}\n}}\n\n'
-                         .format(str(samples[rule][1]), str(samples[rule][0]), call))
-            probe_out += ('\tif(samp_{0} == 0) {{\n'
-                          '\t\t{1}'
-                          '\t}}\n}}\n\n'.format(str(samples[rule][1]), ret))
-        else:
-            # Probe does not need to be sampled
-            probe_in += '\t{0}}}\n\n'.format(call)
-            probe_out += '\t{0}}}\n\n'.format(ret)
+    return kwargs
 
-        # Add probe and return points to the script
-        script += probe_in + probe_out
 
-    return script
+def _pair_rules(probes):
+    """Pairs the rules according to convention:
+     - rule names with '#' serving as a delimiter between two probes, which should be paired as a starting and
+       ending probe
+     - rules with <name>_end or <name>_END are paired with corresponding <name> probes
 
+     :param list probes: the list of probes (as dicts) that should be paired
+     :returns list: probe dictionaries with optionally added 'pair' key containing paired probe name
+    """
+    result = []
+    for probe in probes:
+        # Split the probe definition into pair or probes
+        delim = probe['name'].find('#')
+        if delim != -1:
+            probe['pair'] = probe['name'][delim + 1:]
+            probe['name'] = probe['name'][:delim]
+            result.append(probe)
+        elif probe['name'].endswith('_end') or probe['name'].endswith('_END'):
+            # Skip the end probes
+            continue
```

Comment:
I think the if/elif/else flow can be simplified to just if/elif, by negating the elif condition and including the body of else (this will raise the coverage).